android/app/src/main/java/app/organicmaps/car/screens/bookmarks/BookmarksLoader.java

```diff
@@ -0,0 +1,124 @@
+package app.organicmaps.car.screens.bookmarks;
+
+import android.graphics.drawable.Drawable;
+import android.location.Location;
+import android.text.SpannableStringBuilder;
+import android.text.Spanned;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.car.app.CarContext;
+import androidx.car.app.constraints.ConstraintManager;
+import androidx.car.app.model.CarIcon;
+import androidx.car.app.model.DistanceSpan;
+import androidx.car.app.model.ForegroundCarColorSpan;
+import androidx.car.app.model.ItemList;
+import androidx.car.app.model.Row;
+import androidx.core.graphics.drawable.IconCompat;
+
+import app.organicmaps.R;
+import app.organicmaps.bookmarks.data.BookmarkCategory;
+import app.organicmaps.bookmarks.data.BookmarkInfo;
+import app.organicmaps.bookmarks.data.BookmarkManager;
+import app.organicmaps.bookmarks.data.Icon;
+import app.organicmaps.car.util.Colors;
+import app.organicmaps.car.util.RoutingHelpers;
+import app.organicmaps.location.LocationHelper;
+import app.organicmaps.util.Distance;
+import app.organicmaps.util.Graphics;
+import app.organicmaps.util.concurrency.ThreadPool;
+import app.organicmaps.util.concurrency.UiThread;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+public class BookmarksLoader
+{
+  public interface OnBookmarksLoaded
+  {
+    void onBookmarksLoaded(@NonNull ItemList bookmarks);
+  }
+
+  // The maximum size should be equal to ConstraintManager.CONTENT_LIMIT_TYPE_LIST.
+  // However, having more than 50 items results in android.os.TransactionTooLargeException.
+  // This exception occurs because the data parcel size is too large to be transferred between services.
+  // The primary cause of this issue is the icons. Even though we have the maximum Icon.TYPE_ICONS.length icons,
```

Comment:
Aren't bookmark icons loaded by their id in java? Can that ID be passed instead of an icon? Can icons be loaded here in AA and shared?

Reply:
Not so easy.
We have resId as an icon and color as a background.
We use this data to create a drawable that will be passed to AA (same as on the phone).

> Can icons be loaded here in AA and shared?

I'm not getting it. Could you elaborate?

Reply:
The only place where icons can be loaded from the id is _before_ passing them to AA, right?

Reply:
I'm still not following you. What do you mean by _passing them to AA_?
We are already in AA-related code. Do you mean _passing them to the AA screen_? 
In this case, the answer is yes. You must load data before rendering it on the screen.

There is a possibility to dynamically load icons in AA, but we encounter a limitation due to the requirement to specify the URI for each icon. As we do not store the complete icon, we must instead create an icon dynamically from the resource ID and background color and pass it directly to AA.

Anyway, there is no sense in trying to fix this issue. Because there is no issue in limiting the content at all.
Just imagine yourself sitting in the car and scrolling the list of 100 bookmarks. Even 50 is too much.